ompi/mca/coll/ftagree/coll_ftagree_module.c

```diff
@@ -0,0 +1,180 @@
+ * Copyright (c) 2004-2005 High Performance Computing Center Stuttgart,
+ *                         University of Stuttgart.  All rights reserved.
+ * Copyright (c) 2004-2005 The Regents of the University of California.
+ *                         All rights reserved.
+ * Copyright (c) 2010-2012 Oak Ridge National Labs.  All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "ompi_config.h"
+#include "opal/util/bit_ops.h"
+#include "ompi/mca/coll/coll.h"
+#include "ompi/mca/coll/base/base.h"
+
+#include "coll_ftagree.h"
+#include "coll_ftagree_era.h"
+
+#if OPAL_ENABLE_FT_MPI
+static int
+mca_coll_ftagree_noft(void *contrib,
+                      int dt_count,
+                      struct ompi_datatype_t *dt,
+                      struct ompi_op_t *op,
+                      struct ompi_group_t **group, bool update_grp,
+                      struct ompi_communicator_t* comm,
+                      struct mca_coll_base_module_2_3_0_t *module)
+{
+    return comm->c_coll->coll_allreduce(MPI_IN_PLACE, contrib, dt_count, dt, op,
+                                       comm, comm->c_coll->coll_allreduce_module);
+}
+
+static int
+mca_coll_ftagree_inoft(void *contrib,
+                       int dt_count,
+                       struct ompi_datatype_t *dt,
+                       struct ompi_op_t *op,
+                       struct ompi_group_t **group, bool update_grp,
+                       struct ompi_communicator_t* comm,
+                       ompi_request_t **request,
+                       struct mca_coll_base_module_2_3_0_t *module)
+{
+    return comm->c_coll->coll_iallreduce(MPI_IN_PLACE, contrib, dt_count, dt, op,
+                                        comm, request, comm->c_coll->coll_iallreduce_module);
+}
+
+#endif /* OPAL_ENABLE_FT_MPI */
+
+/*
+ * Initial query function that is invoked during MPI_INIT, allowing
+ * this component to disqualify itself if it doesn't support the
+ * required level of thread support.
+ */
+int
+mca_coll_ftagree_init_query(bool enable_progress_threads,
+                            bool enable_mpi_threads)
+{
+    if( mca_coll_ftagree_algorithm == COLL_FTAGREE_EARLY_RETURNING ) {
+        return mca_coll_ftagree_era_init();
+    }
+
+    return OMPI_SUCCESS;
+}
+
+
+/*
+ * Invoked when there's a new communicator that has been created.
+ * Look at the communicator and decide which set of functions and
+ * priority we want to return.
+ */
+mca_coll_base_module_t *
+mca_coll_ftagree_comm_query(struct ompi_communicator_t *comm,
+                            int *priority)
+{
+    int size;
+    mca_coll_ftagree_module_t *ftagree_module;
+
+    ftagree_module = OBJ_NEW(mca_coll_ftagree_module_t);
+    if (NULL == ftagree_module) return NULL;
+
+    *priority = mca_coll_ftagree_priority;
+
+    /*
+     * Allocate the data that hangs off the communicator
+     * Intercommunicators not currently supported
+     */
+    if( ompi_ftmpi_enabled ) {
```

Comment:
No, we need to populate with a non-null function all fields of the comm module, otherwise we get an error about not being able to instantiate the comm. Our strategy is to  populate with an allreduce based implementation when fault tolerance is not enabled.

Reply:
to clarify, when ENABLE_FT_MPI is not set, we do not even compile ftagree. So this is to manage only the case where it is compiled in, and mpirun --disable-recovery is used.

Reply:
there was no need to clarify, I guess the #ifdef were pretty clear. I was talking about the case when it is compiled in but not enabled for the run. And for that it is still not clear to me what you are referring here to, but I suspect your issue is arising from the CHECK_NULL macro in coll_base_comm_select.c. The non-ft versions of the collectives we add here should go in the base, so that if no FT module is ever loaded that selection will still be valid. Otherwise, if the list of collective specify by the user does not include ftagree, nothing will work.

Reply:
It's done:
1. we set ompi_coll_base_agree_noft in basic
2. if enable_ft_mpi is false, we bail-out immediately from ftagree, thus we keep using allreduce for agree implementation.
3. if enable_ft_mpi is true, era is substituted on the comm (as per normal select logic)

Please mark the conversation resolved if that looks good.